File: misc/triple_integral.c
#include <stdio.h>
#include <stdlib.h>
#define _USE_MATH_DEFINES
#include <math.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

typedef struct {
    double x, y, z;
} Point;

typedef struct {
    double r1, r2, t1, t2, p1, p2;
} Space;

typedef struct {
    int nr, nt, np;
} Resolution;

typedef struct {
    double dr, dt, dp;
    Point *points;
    int num_points;
} Domain;

double delta(double a, double b, int n) {
    return (b - a) / n;
}

void interval(double a, double d, int n, double *result) {
    for (int i = 0; i < n; i++) {
        result[i] = a + d * i;
    }
}

double normr(double r) {
    return r < 0 ? 0 : r;
}

double normt(double t) {
    while (t < 0) t += 2 * M_PI;
    while (t >= 2 * M_PI) t -= 2 * M_PI;
    return t;
}

double normp(double p) {
    if (p < 0) return 0;
    if (p > M_PI) return M_PI;
    return p;
}

Domain createDomain(Space space, Resolution res) {
    Domain domain;
    domain.num_points = res.nr * res.nt * res.np;
    domain.points = (Point *)malloc(domain.num_points * sizeof(Point));
    if (domain.points == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }
    
    domain.dr = delta(space.r1, space.r2, res.nr);
    domain.dt = delta(space.t1, space.t2, res.nt);
    domain.dp = delta(space.p1, space.p2, res.np);
    
    double *rr = (double *)malloc(res.nr * sizeof(double));
    double *tt = (double *)malloc(res.nt * sizeof(double));
    double *pp = (double *)malloc(res.np * sizeof(double));
    
    if (rr == NULL || tt == NULL || pp == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        free(rr);
        free(tt);
        free(pp);
        free(domain.points);
        exit(EXIT_FAILURE);
    }
    
    interval(space.r1, domain.dr, res.nr, rr);
    interval(space.t1, domain.dt, res.nt, tt);
    interval(space.p1, domain.dp, res.np, pp);
    
    int index = 0;
    for (int i = 0; i < res.nr; i++) {
        for (int j = 0; j < res.nt; j++) {
            for (int k = 0; k < res.np; k++) {
                domain.points[index++] = (Point){rr[i], tt[j], pp[k]};
            }
        }
    }
    
    free(rr);
    free(tt);
    free(pp);
    
    return domain;
}

double getVolume(Domain domain, Point p) {
    return p.x * p.x * sin(p.z) * domain.dr * domain.dt * domain.dp;
}

Point *getPoints(Domain domain) {
    return domain.points;
}

double integrate(Domain domain, int (*predicate)(Point), double (*function)(Point)) {
    double sum = 0;
    for (int i = 0; i < domain.num_points; i++) {
        Point p = domain.points[i];
        if (predicate(p)) {
            sum += function(p) * getVolume(domain, p);
        }
    }
    return sum;
}

int predicate(Point p) {
    return 1; // Integrate over all points
}

double function(Point p) {
    return 1;
}

void freeDomain(Domain *domain) {
    free(domain->points);
    domain->points = NULL;
    domain->num_points = 0;
}

int main() {
    Space space = {0, 3, 0, 2 * M_PI, 0, M_PI};
    Resolution res = {100, 200, 200};

    Domain domain = createDomain(space, res);

    double volume = integrate(domain, predicate, function);
    printf("Calculated volume: %f\n", volume);

    freeDomain(&domain);
    return 0;
}

File: misc/triple_integral.py
import numpy as np

class Point:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

class Space:
    def __init__(self, r1, r2, t1, t2, p1, p2):
        self.r1 = r1
        self.r2 = r2
        self.t1 = t1
        self.t2 = t2
        self.p1 = p1
        self.p2 = p2

class Resolution:
    def __init__(self, nr, nt, np):
        self.nr = nr
        self.nt = nt
        self.np = np

class Domain:
    def __init__(self, space, res):
        self.num_points = res.nr * res.nt * res.np
        self.dr = self.delta(space.r1, space.r2, res.nr)
        self.dt = self.delta(space.t1, space.t2, res.nt)
        self.dp = self.delta(space.p1, space.p2, res.np)
        
        rr = self.interval(space.r1, self.dr, res.nr)
        tt = self.interval(space.t1, self.dt, res.nt)
        pp = self.interval(space.p1, self.dp, res.np)
        
        self.points = [Point(r, t, p) for r in rr for t in tt for p in pp]

    def delta(self, a, b, n):
        return (b - a) / n

    def interval(self, a, d, n):
        return [a + d * i for i in range(n)]

def get_volume(domain, p):
    return p.x * p.x * np.sin(p.z) * domain.dr * domain.dt * domain.dp

def integrate(domain, predicate, function):
    sum = 0
    for p in domain.points:
        if predicate(p):
            sum += function(p) * get_volume(domain, p)
    return sum

def predicate(p):
    return True  # Integrate over all points

def function(p):
    return 1

def main():
    space = Space(0, 1, 0, 3 * np.pi / 2, 0, np.pi/6)
    res = Resolution(100, 200, 200)

    domain = Domain(space, res)

    volume = integrate(domain, predicate, function)
    print(f"Calculated volume: {volume}")

if __name__ == "__main__":
    main()